paddle/framework/tensor.cc

```diff
@@ -0,0 +1,95 @@
+#include "paddle/framework/tensor.h"
+
+namespace paddle {
+namespace framework {
+
+template <typename T>
+const T* Tensor::Data() const {
+  PADDLE_ASSERT(holder_ != nullptr);
+  PADDLE_ASSERT(holder_->Place() == place_);
+  PADDLE_ASSERT(holder_->Size() >= product(dims_) * sizeof(T));
+  return static_cast<const T*>(holder->Ptr());
+}
+
+bool Tensor::NeedReset() {
+  return (holder_ == nullptr || holder_->Place() != place_ ||
+          holder_->Size() < product(dims_) * sizeof(T));
+}
+
+template <typename T, typename = std::enable_if<std::is_pod<T>::value>::type>
+T* Tensor::MutableData() {
+  if (NeedReset()) {
+    holder_.reset(new PlaceholderImpl(place_, product(dims_) * sizeof(T)));
+  }
+  return static_cast<T*>(holder_->Ptr());
+}
+
+template <typename T, typename = std::enable_if<std::is_pod<T>::value>::type>
+T* Tensor::MutableData(const DDim& dims) {
+  dims_ = dims;
+  return MutableData<T>();
+}
+
+template <typename T, typename = std::enable_if<std::is_pod<T>::value>::type>
+T* Tensor::MutableData(const DDim& dims, const Place& place) {
+  dims_ = dims;
+  place_ = place;
+  return MutableData<T>();
+}
+
+int Tensor::Rank() const { return arity(dims_); }
+
+int Tensor::Numel() const { return product(dims_); }
+
+void Tensor::Resize(const DDim& dims) {
+  dims_ = dims;
+  return;
+}
+
+void Tensor::Reshape(const DDim& dims) {
+  if (product(dims) != product(dims_)) {
+    // TODO: error: "Reshape() can not change tensor's numel".
```

Comment:
Why can not reshape change tensor's numel?

In each mini-batch training, each output of Op should be `Reshape`, because the shape of input tensor could be changed.